Conflicting warnings when building an uberjar with deprecated property `quarkus.package.type=uber-jar`

When building an uberjar with `quarkus.package.type=uber-jar`, we end up with the following warnings:

```
[WARNING] [io.quarkus.deployment.configuration] Configuration property 'quarkus.package.type' has been deprecated and replaced by: [quarkus.package.jar.enabled, quarkus.package.jar.type, quarkus.native.enabled, quarkus.native.sources-only]
[WARNING] [io.quarkus.config] Unrecognized configuration key "quarkus.package.type" was provided; it will be ignored; verify that the dependency extension for this configuration is set or that you did not make a typo
[INFO] [io.quarkus.deployment.pkg.steps.JarResultBuildStep] Building uber jar: /home/gsmet/git/monitor-artifact-publication-action/target/monitor-artifact-publication-action-999-SNAPSHOT-runner.jar
```

Given an uberjar is actually built, I think the second warning is not ideal as it's saying the configuration will be ignored, which is untrue.
I haven't checked exactly how this change was implemented but maybe we should keep the old deprecated config around to avoid the second warning?

/cc @dmlloyd

I didn't get the second warning, but I did get this error:

```
[INFO] --- quarkus:999-SNAPSHOT:build (default) @ monitor-artifact-publication-action ---
[WARNING] [io.quarkus.deployment.configuration] Configuration property 'quarkus.package.type' has been deprecated and replaced by: [quarkus.package.jar.enabled, quarkus.package.jar.type, quarkus.native.enabled, quarkus.native.sources-only]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  2.117 s
[INFO] Finished at: 2024-04-12T11:05:09-05:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal io.quarkus.platform:quarkus-maven-plugin:999-SNAPSHOT:build (default) on project monitor-artifact-publication-action: Failed to build quarkus application: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
[ERROR] 	[error]: Build step io.quarkus.rest.client.reactive.deployment.RestClientReactiveProcessor#announceFeature threw an exception: java.lang.NoSuchFieldError: REST_CLIENT_REACTIVE
[ERROR] 	at io.quarkus.rest.client.reactive.deployment.RestClientReactiveProcessor.announceFeature(RestClientReactiveProcessor.java:140)
[ERROR] 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[ERROR] 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
[ERROR] 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[ERROR] 	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
[ERROR] 	at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:849)
[ERROR] 	at io.quarkus.builder.BuildContext.run(BuildContext.java:256)
[ERROR] 	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
[ERROR] 	at org.jboss.threads.EnhancedQueueExecutor$Task.doRunWith(EnhancedQueueExecutor.java:2516)
[ERROR] 	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2495)
[ERROR] 	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1521)
[ERROR] 	at java.base/java.lang.Thread.run(Thread.java:840)
[ERROR] 	at org.jboss.threads.JBossThread.run(JBossThread.java:483)
[ERROR] -> [Help 1]
```

This is with a fresh Quarkus build. Did I miss something?

--- a/core/deployment/src/main/java/io/quarkus/deployment/configuration/BuildTimeConfigurationReader.java
+++ b/core/deployment/src/main/java/io/quarkus/deployment/configuration/BuildTimeConfigurationReader.java
@@ -1086,6 +1086,7 @@ public final class BuildTimeConfigurationReader {
             builder.getSources().clear();
             builder.getSourceProviders().clear();
             builder.setAddDefaultSources(false)
+                    .withInterceptors(ConfigCompatibility.FrontEnd.nonLoggingInstance(), ConfigCompatibility.BackEnd.instance())
                     .addDiscoveredCustomizers()
                     .withProfiles(config.getProfiles())
                     .withSources(sourceProperties);
@@ -1099,6 +1100,7 @@ public final class BuildTimeConfigurationReader {
             builder.getSources().clear();
             builder.getSourceProviders().clear();
             builder.setAddDefaultSources(false)
+                    .withInterceptors(ConfigCompatibility.FrontEnd.nonLoggingInstance(), ConfigCompatibility.BackEnd.instance())
                     .addDiscoveredCustomizers()
                     .withSources(sourceProperties)
                     .withSources(new MapBackedConfigSource(

--- a/core/deployment/src/main/java/io/quarkus/deployment/configuration/ConfigCompatibility.java
+++ b/core/deployment/src/main/java/io/quarkus/deployment/configuration/ConfigCompatibility.java
@@ -114,9 +114,13 @@ public final class ConfigCompatibility {
         @Serial
         private static final long serialVersionUID = -3438497970389074611L;
 
-        private static final FrontEnd instance = new FrontEnd();
+        private static final FrontEnd instance = new FrontEnd(true);
+        private static final FrontEnd nonLoggingInstance = new FrontEnd(false);
 
-        private FrontEnd() {
+        private final boolean logging;
+
+        private FrontEnd(final boolean logging) {
+            this.logging = logging;
         }
 
         public ConfigValue getValue(final ConfigSourceInterceptorContext context, final String name) {
@@ -155,11 +159,13 @@ public final class ConfigCompatibility {
                         // get the replacement names
                         List<String> list = fn.apply(context, new NameIterator(next));
                         subIter = list.iterator();
-                        // todo: print these warnings when mapping the configuration so they cannot appear more than once
-                        if (list.isEmpty()) {
-                            log.warnf("Configuration property '%s' has been deprecated and will be ignored", next);
-                        } else {
-                            log.warnf("Configuration property '%s' has been deprecated and replaced by: %s", next, list);
+                        if (logging) {
+                            // todo: print these warnings when mapping the configuration so they cannot appear more than once
+                            if (list.isEmpty()) {
+                                log.warnf("Configuration property '%s' has been deprecated and will be ignored", next);
+                            } else {
+                                log.warnf("Configuration property '%s' has been deprecated and replaced by: %s", next, list);
+                            }
                         }
                     }
                     return true;
@@ -178,6 +184,10 @@ public final class ConfigCompatibility {
 
         public static FrontEnd instance() {
             return instance;
+        }
+
+        public static FrontEnd nonLoggingInstance() {
+            return nonLoggingInstance;
         }
     }
 

